pkg/R/sparkR.R

```diff
@@ -121,45 +119,38 @@ sparkR.init <- function(
   if (sparkRExistingPort != "") {
     sparkRBackendPort <- sparkRExistingPort
   } else {
+    # TODO: test the random port and retry, or use httpuv:::startServer
+    callbackPort = sample(40000, 1) + 10000
     if (Sys.getenv("SPARKR_USE_SPARK_SUBMIT", "") == "") {
       launchBackend(classPath = cp,
                     mainClass = "edu.berkeley.cs.amplab.sparkr.SparkRBackend",
-                    args = as.character(sparkRBackendPort),
+                    args = as.character(callbackPort),
                     javaOpts = paste("-Xmx", sparkMem, sep = ""))
     } else {
       # TODO: We should deprecate sparkJars and ask users to add it to the
       # command line (using --jars) which is picked up by SparkSubmit
       launchBackendSparkSubmit(
           mainClass = "edu.berkeley.cs.amplab.sparkr.SparkRBackend",
-          args = as.character(sparkRBackendPort),
+          args = as.character(callbackPort),
           appJar = .sparkREnv$assemblyJarPath,
           sparkHome = sparkHome,
           sparkSubmitOpts = Sys.getenv("SPARKR_SUBMIT_ARGS", ""))
     }
+    sock <- socketConnection(port = callbackPort, server = TRUE, open = 'rb',
```

Comment:
Hmm what about this solution -- We pass in a temp file name as the first argument to SparkRRunner which writes out the port number to that file. We can use `tempfile`[1] to get a unique file name and the one complexity is that we might need to retry the file read.

I'm just trying to see if we can avoid the risk of picking a port that is already used in `socketConnection` or of the JVM race condition etc.


[1] https://stat.ethz.ch/R-manual/R-patched/library/base/html/tempfile.html